linked_list/linked_list.py

```diff
@@ -16,64 +19,152 @@ def __init__(self):
+
+        current = self.head
+        count = 0
+
+        while current:
+            if count == index:
+                return current.value
+            count += 1
+            current = current.next
+        return None 
+
+        
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)
     def get_last(self):
-        pass
+        if not self.head:
+            return None
+        current = self.head 
+        while current.next:
+            current = current.next
+        return current.value 
+        # return self.tail.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)
     def add_last(self, value):
-        pass
+        if not self.head:
+            self.add_first(value)
+        else: 
+            new_node = Node(value)
+            current = self.head
+            while current.next: 
+                current = current.next
+            current.next = new_node 
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if not self.head: 
+            return None 
+        current = self.head
+        max = current.value
+
+        while current: 
+            if max < current.value: 
+                max = current.value 
+            current = current.next
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)
     def delete(self, value):
-        pass
+        if not self.head: 
+            return 
+        # current = self.head 
+        elif current.value == value:
+            self.head = current.next
+            return True 
+        else:
+            current = self.head 
+            while current.next and current.next.value != value: 
+                current = current.next 
+            current.next = current.next.next
+            # if self.head:
+            #     self.head.previous = None
+            # if not self.head:
+            #     self.tail = None    
+            # return None 
+
+        #   previous = current
+
+        # while current.next:
+        #     if current.next.value == value:
+        #         current.next = current.next.next
+        #         if current.next:
+        #             current.next.previous = current
+        #         if not current.next:
+        #             self.tail = current
+
+        #         return None
+         
+
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)            

```

Comment:
🪐 Space complexity will be O(n) here since you are creating the list `helper_list` which will end up holding all of the nodes in the linked list